### ISSUE TITLE
Implement io.Seeker for decrypting with random access 

### ISSUE THREAD

**Issue by aead** [minio/sio#13]
Decrypting (and maybe also encrypting) data streams using random access is an important use case.
Therefore the decrypting io.Reader and io.Writer (and maybe also the encrypting counterparts) should implement io.Seeker.

Seeking can be done from the beginning of the stream (io.SeekStart) or from the current position (io.SeekCurrent) but not from the end (io.SeekEnd) because the size of the stream is usually not known.

Seeking within an encrypted stream from io.SeekStart requires:
 - Moving the position of the underlying io.Reader to the `k`-th package - `k` can be computed by `(offset / payload_size) * (payload_size + 32)`.
 - Adjusting the expected sequence_number to `k`. The `k`-th package is decrypted so the expected sequence-number must be equal to `k`
- If the seeking `offset` is not a multiple of the payload_size the next `mod := offset % payload_size` bytes must be extracted and discarded (but verified).

Seeking within an encrypted stream from io.SeekCurrent requires:
 - Moving the position of the underlying io.Reader to the `k`-th package - `k` can be computed by `sequence_number + (offset / payload_size) * (payload_size + 32)`.
 - Adjusting the expected sequence_number to `k`. The `k`-th package is decrypted so the expected sequence-number must be equal to `k`
- If the seeking `offset` is not a multiple of the payload_size the next `mod := current_offset + (offset % payload_size)` bytes must be extracted and discarded (but verified). The `current_offset` is the current offset of the reader within the currently package. An important corner case is that `mod` can be negative. In this case `k = k-1` and `mod = payload_size - |mod|`

## Seeking from the user perspective
The most important rule for users of the sio library is: 
**Seeking within an encrypted stream is only valid if all packages (except for the last one) have the same package size.**

This means that the length of all packages - except the length of the last package - **must** be the same. Further if a custom package size is used (the `Config.PayloadSize` field is set manually for encryption) it is **required** that the `Config.PayloadSize` is set to the same value for decryption. Notice that this is not required to decrypt a stream but to compute the 'correct' package and sequence number.
In all cases where the packages (except for the last one) have not same package size or the wrong payload size is used the seeking may succeed but the decryption will **always** fail. 

It is **not** possible to seek within an encrypted stream (using io.Seeker) if the size of packages within one stream is not the same - e.g. If the payload size of package `1` is 32 `KB`, package `2` is 33 `KB`, 
package `3` is 55 `KB`, ... a.s.o.
In this case the user must set the correct configuration values and implement the seeking functionality on top of the sio library manually.

**Comment by aead** [comment]
## Proposed behavior
It should be possible to seek back and forward in an encrypted stream if:
 - **All packages of the encrypted stream except for the last package have the same package size.**
   This is the default behavior during encryption.
 - **The payload size during decryption is the same as during encryption.** This is also the default
   behavior.

Currently only decryption should support seeking. As far as I can see there is no important use case for random access encryption. Even worse it may lead to additional pitfalls which may result in wrongly encrypted data streams.

Further it's probably okay to implement io.Seeker only for the io.Reader returned by `DecryptReader`.
I think it's the mos common use case to seek within a encrypted stream during reading not during writing. 

## Implementation details
Implementing io.Seeker for the decrypting reader should only support `io.SeekStart` and `io.SeekCurrent`. The problem when trying to seek from the end of a stream is that the size/length of the entire stream is not known most of the time. Seeking from `io.SeekCurrent` can be implemented by computing the current position within the stream, adding this position to the provided offset and doing a seek from the start.

The decrypting reader can only provide a constant time seek operation if the underlying io.Reader implements io.Seeker. Unfortunately this cannot be checked statically or properly during the decrypting reader initialization. Therefore the seeking may return an error (at runtime on the first call to `Seek`) when the underlying io.Reader does not implement io.Seeker.

**Comment by AudriusButkevicius** [comment]
I think io.WriterAt and io.ReaderAt would be the right way to go in terms of implementation.
Seeking can be implemented using those.

Encryption supporting WriteAt or seeking would be very useful for syncthing as we fetch data in random order. I know supporting WriteAt for encryption is problematic, yet I suggest the implementation to require WriteAt's to happen at package boundaries, which makes the hard problem go away.

**Comment by aead** [comment]
>I think io.WriterAt and io.ReaderAt would be the right way to go in terms of implementation.
Seeking can be implemented using those.

Yes, I agree here - that this is probably the way to go.

>Encryption supporting WriteAt or seeking would be very useful for syncthing as we fetch data in random order. I know supporting WriteAt for encryption is problematic, yet I suggest the implementation to require WriteAt's to happen at package boundaries with package sizes, which makes the hard problem go away.

If I understood the [syncthing protocol specification](https://docs.syncthing.net/specs/bep-v1.html#introduction-and-definitions) you split an arbitrary data stream into 128K blocks. Each block can be fetched separately depending on the diff. state. 
A few questions: 
1. What would happen if data blocks are appended to an existing blob?
2. What would happen if data blocks are removed from an existing blob?
   If this is a valid operation how should the encryption scheme differentiate between a
   valid user operation (the user removed some data) and an attack?
3. If data is synced as complete blob - e.g I modify my file and the whole file is synced - why would random access writes be required? 
4. If data is synced as individual blocks (not as complete blob) and a random access write is used - an attacker can always XOR the new with the old block an gets the XOR of both plaintexts because the keys would be the same ( if the keys are not equal you would have two blocks of the same file with different keys)?

**Comment by AudriusButkevicius** [comment]
So the idea would be that users only modify plain text files. Any modification to a plain text file would result in fully rerunning DARE on the file and advertising new encrypted blocks to others. We would then diff on the receiving end what blocks have changed only download the blocks that changed. DARE wise we'd optionally reuse the nonce for a given file to enable reusing blocks from previous version of the encrypted file (trading off security for efficiency by using nonce more like a salt). In syncthing context this might end up as two blocks changing, as for example in the new DARE spec, the final flag moves from old last block to the new last block. (Or only last block changing if the append is within the same sub-block size blockmu)

Any modifications to encrypted files should result in fatal errors. I know current iteration of DARE does not prevent truncate and append attacks. I guess we can live with that for now while DARE improves.

I guess this answers 1 and 2.

3. Data is initially synced as a complete blob (yet fetching blocks for the blob in random order), but any changes after that are incremental as we diff block between the old file (the file that we have) and the new file that is advertised, and fetch only changed blocks.  If this file is a VM image, changes in the middle of the file are possible, and we'd hope to only transfer blocks in the middle.

4. It's synced as random blocks that build up the blob. So we'd still apply DARE at a file/whole blob level, yet write out the DARE packages/syncthing blocks out of order. Syncthing tries to be torrent like, as in peer A advertises some file, peers B and C start downloading blocks from A in random order and as they are downloaded on B or C, they become available to others. So any block downloaded by B, suddenly becomes available to C in case it decides to ask B for that block instead of A (and it will to utilize bandwidth better). If we were to fetch blocks sequentially, it's likely that B and C will always have the same blocks as they download the file in the same order, preventing using bandwidth between B and C.

I am not sure I understand any XOR attack. If you XOR the old block and the new block (which is potentially equal to the old block plus some few new bytes at the end in terms of plain text), I am not sure what that would give you? I assume the stream cipher would result in the same stream in both blocks up to the point where new bytes are appended to the new block? 

If the plain text behind the block actually changes, then the XOR would give you garbage?

**Comment by aead** [comment]
@AudriusButkevicius Okay, thanks for the clarification.
So I see the following problem based on your description:

Let's assume a user A initially syncs a file `f` consisting of 3 DARE packages encrypted with a secret key `k` - for syncthing 2 128K blocks (I guess the last block is padded somehow). Now the file is distributed two another node B. A modifies the data of the 2. package. Now there are four options: 
1. Encrypt the whole plaintext file `f` again with the same key `k`
   Than only the first 128K block has changed and must be synced. However an attacker
   can XOR original 2. package with the new 2. package and receives:
   <code>P<sub>2</sub> XOR P'<sub>2</sub> = old_plaintext XOR updated_plaintext</code>  
  This is the case because the keystream for the 2. package would be the same. You basically
   would reuse a key **and** a nonce. That breaks security. 
2. Encrypt the whole plaintext file `f` with a different key `k'`
  Than the whole file has changed and must be synced. But this would be secure as far as I can see.
3. Encrypt only the 2. package (upper 1. 128K block) with the same key `k`
   Same problem like 1.
4. Encrypt only the 2. package (upper 1. 128K block) with a different key `k'`
   The updated encrypted file would consist of a 2 packages encrypted with `k` and 1 package 
   encrypted with `k'`. That's not a valid encrypted DARE blob. You would not be able to decrypt.

The main problem I see here is that modifying encrypted data (partially) would be the default use case. So you need a encryption scheme / protocol which allows you to modify parts of your encrypted data securely without encrypting the whole plaintext again.

**Comment by AudriusButkevicius** [comment]
Right, thanks for having this discussion with me, I had to remind myself how stream ciphers work which points out that my idea is completely flawed.
I don't think DARE works for us, as we either have to derive a new key every time a file changes., or have a nonce per block and change the nonce every time the block changes.

I was hoping to do encryption completely transparently at the filesystem level, so syncthing would need minimal changes, yet I now understand that this is impossible, as we have to be able to distinguish reads for scanning (which need to change key/nonce) and reads for requests (which need to use the old nonce/key). 

I guess having an IV/nonce per block and changing that nonce/IV every time the underlying block changes would work.

**Comment by aead** [comment]
Sure, no problem. 
>I was hoping to do encryption completely transparently at the filesystem level, so syncthing would need minimal changes, yet I now understand that this is impossible, as we have to be able to distinguish reads for scanning (which need to change key/nonce) and reads for requests (which need to use the old nonce/key).
>
>I guess having an IV/nonce per block and changing that nonce/IV every time the underlying block changes would work.

Yes, since syncthing already splits the data into fixed size blocks confidentiality and integrity of blocks can be achieved easily - an AEAD cipher like AES-GCM or ChaCha20Poly1305 should be sufficient.
However if you also try to achieve integrity not just for single blocks but also for the complete blob - which I would recommend -  than a slightly more complex scheme is required.

Nevertheless let me know if you need a critical review 😉

**Comment by AudriusButkevicius** [comment]
Well, if I have a header like DARE, with block index within the blob, payload size and whether it's final or not, I think I should be good for a given blob. I guess I can still swap first blocks of different files if they use the same key, but I can just use pbkdf and have yet another nonce that is used to derive the key.

**Comment by aead** [comment]
Right, but you must also ensure that a "new" block cannot be replaced by an "old" block. For example the user changes file `f` in such a way that a specific block `b` changes to `b'` and must be synced. "Just" using a DARE-like scheme would not prevent an adversary from replaced `b'` with `b` again during sync.

**Comment by AudriusButkevicius** [comment]
Worst thing here is corrupting or reverting files, which is bad but not tragic.
I can think of how to prevent block replacement, yet I can't think of a way of preventing replacing whole blobs. Swap file A with file B for example, or older version of file A.

**Comment by aead** [comment]
Swapping file A with file B can be prevented by using unique keys per file.

Replacing blocks of file A with older versions can be prevented with an additional counter. For example a block version counter. So initially set to zero it gets incremented after every change to the block. A receiver would reject any block that has a lower or the same "block version counter" than the receiver's block. This way an attacker should not be able to replay old blocks.

**Comment by Xeoncross** [comment]
> Seeking within an encrypted stream is only valid if all packages (except for the last one) have the same package size.

I must be missing something, but I thought that you could read each header, skip the N-bytes long body, then read the next header (and so on).

**Comment by harshavardhana** [comment]
> I must be missing something, but I thought that you could read each header, skip the N-bytes long body, then read the next header (and so on).

@Xeoncross with DARE 2.0 this is quite possible.
